OrbitClientGgp: project structure and basic capture flow

kisenshi opened the pull request:
### Create OrbitClientGgp to manage a capture request from the ggp instance

Main files and the structure of the project has been created. Basic flow is working: initial set up, starts capture, captures for a period of time and stops the capture.

Capture data is not collected or handled yet; this pull request is intended to include the project in Orbit so there is not a huge list of commits later. Also it avoids being 'silently' broken in updates related to Capture, etc. 

The options related to the capture cycle must be provided as an argument to OrbitClientGgp. Some values are given a default value but others are required to be provided by the user:

* pid. (Mandatory) Id of the process to capture.
* grpc_port. (Optional) Required for gRPC, by default is 44765.
* sampling_rate. (Optional) Required for CaptureClient to work, by default is 1000.
* frame_pointer_unwinding. (Optional) Required for CaptureClient to work.
* capture_length. (Optional) Sets the length of the capture in seconds, by default is 10

The new project contains the following files:

+ main.cpp Contains the linear flow and handles the different classes instantiation and requests
+ ClientGgp.h/cc Contains methods that handle the gRPC client and capture requests. Extends and implements CaptureListener (although it does not handle the calls yet).
+ ClientGgpOptions.h  Struct containing the options needed for the client initialisation.

Also a Run_client_ssh.ps1 script has been created to facilitate the deployment and test of OrbitClientGgp.

beckerhe commented on OrbitClientGgp/ClientGgp.cpp:
```diff
@@ -0,0 +1,134 @@
+
+void ClientGgp::OnTimer(const orbit_client_protos::TimerInfo& timer_info) {
+  (void)timer_info;
+  //LOG("OnTimer called");
```
Can you remove these LOG-statements if they are not needed?

kisenshi replied:
Done

beckerhe commented on OrbitClientGgp/CMakeLists.txt:
```diff
@@ -0,0 +1,25 @@
+    OrbitCaptureClient
+    OrbitProtos)
+
+#strip_symbols(OrbitClientGgp)
```
Can you remove that line if not needed? Is there a reason why you commented that?

kisenshi replied:
Done. I saw it was in the OrbitService so I added and commented out here in case I needed as well. I don't think I do in the end, so I removed it

beckerhe commented on run_client_ssh.ps1:
```diff
@@ -0,0 +1,59 @@
+# Connects to the instance and starts OrbitClientGgp. To also deploy OrbitClientGgp,
```
I think a license header is missing here.

kisenshi replied:
You are right, added!

beckerhe commented on OrbitClientGgp/main.cpp:
```diff
@@ -0,0 +1,73 @@
+
+  LOG("Go to sleep");
+  // Captures for the period of time requested
+  sleep(capture_length);
```
You're probably calling the POSIX-function `sleep` here, aren't you? Can you replace that with `absl::SleepFor` which makes it more obvious what unit the argument has and which is more portable?

kisenshi replied:
Makes sense, I have updated it, thanks for the suggestion!

beckerhe commented on OrbitClientGgp/ClientGgp.cpp:
```diff
@@ -0,0 +1,134 @@
+#include "Capture.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/Result.h"
+
```
We should try to explicitly include the headers needed here.

- `std::numeric_limits` needs `<limits>`.
- `std::make_unique` needs `<memory>`. 
- `int32_t` needs `<cstdint>`.
- `std::map` needs `<map>`.
- `std::string` needs `<string>`.

kisenshi replied:
Done. I was trying to avoid including headers that are already included by including others; but maybe it makes sense for these, so I have included them.

beckerhe replied:
Yeah, that's always a thing for debate what headers to include. The Google C++ style guide is pretty clear here though: We should "include all headers defining symbols we rely upon".

Thanks!

beckerhe commented on OrbitClientGgp/ClientGgp.h:
```diff
@@ -0,0 +1,52 @@
+//#include "Callstack.h"
+//#include "EventBuffer.h"
+//#include "KeyAndString.h"
+//#include "ScopeTimer.h"
```
And add those that are needed for this header? I see `<cstdint>`, `<string>`, and `<memory>`.

kisenshi replied:
Done and added as well

danielfenner commented on OrbitClientGgp/main.cpp:
```diff
@@ -0,0 +1,73 @@
+
+int main(int argc, char** argv) {
+  absl::ParseCommandLine(argc, argv);
+  uint64_t grpc_port_ = absl::GetFlag(FLAGS_grpc_port);
```
Remove underscore at the end - that's for member variables only.

kisenshi replied:
True, I didn't realise, thanks!

danielfenner commented on OrbitClientGgp/main.cpp:
```diff
@@ -0,0 +1,73 @@
+  uint32_t capture_length = absl::GetFlag(FLAGS_capture_length);
+  
+  if (!absl::GetFlag(FLAGS_pid)) {
+    ERROR("pid to capture not provided; set using -pid");
```
You could just use FATAL here.

kisenshi replied:
Fixed

danielfenner commented on OrbitClientGgp/main.cpp:
```diff
@@ -0,0 +1,73 @@
+
+  client_ggp.StopCapture();
+
+  LOG("Shuting down the thread pool");
```
Shutting

kisenshi replied:
I updated the comment because of the use of thread instead

kisenshi commented on OrbitClientGgp/main.cpp:
```diff
@@ -0,0 +1,73 @@
+  std::unique_ptr<ThreadPool> thread_pool = ThreadPool::Create(4 /*min_size*/, 256 /*max_size*/, absl::Seconds(1));
+
+  // The request is done in a separate thread to avoid blocking main()
+  thread_pool->Schedule([&]() {
```
This makes a lot of sense, I prefer doing this then. Thanks!

kisenshi commented on OrbitClientGgp/main.cpp:
```diff
@@ -0,0 +1,73 @@
+#include "absl/flags/parse.h"
+//#include "absl/flags/usage.h"
+//#include "absl/flags/usage_config.h"
+//#include "absl/strings/str_format.h"
```
Done, I forgot I had commented them out

kisenshi commented on OrbitClientGgp/ClientGgp.cpp:
```diff
@@ -0,0 +1,134 @@
+}
+
+void ClientGgp::OnTimer(const orbit_client_protos::TimerInfo& timer_info) {
+  (void)timer_info;
```
Didn't know this; I updated it, thanks for the tip!

kisenshi commented:
> Hi Cris,
> I hope you don't mind that I added some minor comments. It looks really good! :-)
> 
> But someone else who is more familiar with what you're trying to do should also have a look. Supposedly Daniel. Thanks.

Hi, 

All feedback is welcome so thanks a lot for taking the time to take a look at the code :)

I have gone through the comments and suggestions and updated the code accordingly. Thanks!

kisenshi commented:
> Looks good! There is really only some nit picking.

I have updated the code with the suggestions, thanks!

> Besides the points below: maybe have a look into minimize use of vertical whitespace. The style guide asks for as little empty lines as possible. This is not a fixed rule thought, use your best judgement.

Do you mean I should include less empty lines between code? I can take a look and update it. I like including them because I think it helps readability and the separation between different concepts but I can update it to follow the guidelines

kisenshi commented:
@danielfenner I have removed some of the vertical whitespace as suggested to follow the guidelines. I think the code is still clear like this. Thanks!

kisenshi commented:
The code will need to be updated as a result of [PR #980](https://github.com/google/orbit/pull/980) which changes Capture and includes the thread creation inside OrbitClient, so I will wait for that to be merged and update the code accordingly. It will need new review when done.

danielfenner commented:
> Do you mean I should include less empty lines between code? I can take a look and update it. I like including them because I think it helps readability and the separation between different concepts but I can update it to follow the guidelines

Yes exactly, the java style guide encourages exactly this - use empty lines as paragraphs. The c++ style guide asks for the opposite. As I wrote above, just using your own judgement is totaly fine.

kisenshi commented on OrbitClientGgp/main.cpp:
```diff
@@ -0,0 +1,62 @@
+  // The request is done in a separate thread to avoid blocking main()
+  // It is needed to provide a thread pool
+  std::unique_ptr<ThreadPool> thread_pool =
+      ThreadPool::Create(1, 1, absl::Seconds(1));
```
I am not sure about this initialisation, I only need one thread , is this the right way to do it (We need to use ThreadPool because is the one used in CaptureClient)?

danielfenner replied:
I think this is just fine as is; as you wrote you need the thread pool since it is in the signature of StartCapture.

kisenshi commented:
I have updated the code to integrate the changes related to Capture and CaptureClient

danielfenner commented on OrbitClientGgp/ClientGgp.h:
```diff
@@ -0,0 +1,47 @@
+#include "ClientGgpOptions.h"
+#include "OrbitCaptureClient/CaptureClient.h"
+#include "OrbitCaptureClient/CaptureListener.h"
+#include "OrbitProcess.h"  // remove if including Capture.h in here
```
Is this comment outdated? At least I don't get it.

kisenshi replied:
It was a note to myself when I was figuring out what it was needed to include and completely forgot to remove it, will update it, thanks!

danielfenner commented on OrbitClientGgp/ClientGgp.h:
```diff
@@ -0,0 +1,47 @@
+  std::unique_ptr<CaptureClient> capture_client_;
+
+  void InitCapture();
+  std::shared_ptr<Process> GetOrbitProcessByPid(int32_t pid);
```
I think this can be moved into the implementation entirely (e.g. into a free function in the unnamed namespace). No need to hae it in the header.

kisenshi replied:
Done, thanks!

kisenshi commented:
Including OrbitClientGgp into master
